### PR TITLE
Symbolic Cholesky factorization

### PR DESCRIPTION
**Pull request by upsj**
This is a WIP PR I will use to collect the finalized parts of my Cholesky implementation

### CODE REVIEW

File: core/test/components/disjoint_sets.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+
+int popcount(int i)
+{
+    int count{};
+    while (i) {
+        count += i & 1;
+        i >>= 1;
+    }
+    return count;
```
Comment (upsj) [inline core/test/components/disjoint_sets.cpp:133]:
```suggestion
    return std::bitset<32>(i).count();
```

File: core/test/factorization/elimination_forest.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                        I<index_type>({0, 1, 3, 4, 6, 5, 7, 8, 2, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder,
+                        I<index_type>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.inv_postorder,
+                        I<index_type>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder_parents,
+                        I<index_type>({2, 2, 10, 4, 5, 9, 7, 9, 9, 10}));
+}
+
+
```
Comment (fritzgoebel) [inline core/test/factorization/elimination_forest.cpp:141]: In the test cases, the postorder array is self-inverted for some reason. I'd suggest adding this test where this is not the case:

```suggestion

TYPED_TEST(EliminationForest, WorksForPostOrderNotSelfInverted)
{
    using matrix_type = typename TestFixture::matrix_type;
    using index_type = typename TestFixture::index_type;
    auto mtx = gko::initialize<typename TestFixture::matrix_type>(
        {
            {1, 0, 1, 0, 0, 0, 0, 0, 0, 0},
            {0, 1, 0, 0, 1, 0, 0, 0, 0, 0},
            {1, 0, 1, 0, 0, 0, 1, 0, 0, 0},
            {0, 0, 0, 1, 0, 0, 0, 0, 1, 0},
            {0, 1, 0, 0, 1, 1, 0, 0, 0, 0},
            {0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
            {0, 0, 1, 0, 0, 1, 1, 1, 0, 0},
            {0, 0, 0, 0, 0, 0, 1, 1, 1, 0},
            {0, 0, 0, 1, 0, 0, 0, 1, 1, 1},
            {0, 0, 0, 0, 0, 0, 1, 0, 1, 1},
        },
        this->ref);

    auto forest = gko::factorization::compute_elim_forest(mtx.get());

    GKO_ASSERT_ARRAY_EQ(forest.parents,
                        I<index_type>({2, 4, 6, 8, 5, 6, 7, 8, 9, 10}));
    GKO_ASSERT_ARRAY_EQ(forest.child_ptrs,
                        I<index_type>({0, 0, 0, 1, 1, 2, 3, 5, 6, 8, 9, 10}));
    GKO_ASSERT_ARRAY_EQ(forest.children,
                        I<index_type>({0, 1, 4, 2, 5, 6, 3, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.postorder,
                        I<index_type>({3, 0, 2, 1, 4, 5, 6, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.inv_postorder,
                        I<index_type>({1, 3, 2, 0, 4, 5, 6, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.postorder_parents,
                        I<index_type>({8, 2, 6, 4, 5, 6, 7, 8, 9, 10}));
}


```

File: cuda/factorization/cholesky_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        size_type buffer_size{};
+        cusparse::csrsort_buffer_size(handle, num_rows, num_rows, mtx_nnz,
+                                      row_ptrs, postorder_cols, buffer_size);
+        Array<char> buffer_array{exec, buffer_size};
+        auto buffer = buffer_array.get_data();
+        cusparse::csrsort(handle, num_rows, num_rows, mtx_nnz, descr, row_ptrs,
+                          postorder_cols, permutation, buffer);
+        cusparse::destroy(descr);
+    }
+    // count nonzeros in L
```
Comment (fritzgoebel) [inline cuda/factorization/cholesky_kernels.cu:112]:
```suggestion
    // count nonzeros per row of L
```

File: cuda/factorization/cholesky_kernels.cu [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            num_rows, cols, row_ptrs, inv_postorder, postorder_cols,
+            lower_ends);
+    }
+    // sort postorder_cols inside rows
+    {
+        const auto handle = exec->get_cusparse_handle();
+        auto descr = cusparse::create_mat_descr();
+        Array<IndexType> permutation_array(exec, mtx_nnz);
+        auto permutation = permutation_array.get_data();
+        components::fill_seq_array(exec, permutation, mtx_nnz);
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/factorization/cholesky_kernels.cu:102]: Mainly trying to avoid errors about uninitialized data here, but yes, that is totally correct.

**Review by fritzgoebel — APPROVED** [review]

**Review by vasilisge0 — APPROVED** [review]

**Pull request merged**